File: ios/PhiStudioIOS/PhiStudioIOS/Core/ChartDocument.swift
import Foundation

public struct ChartDocument: Hashable, Sendable {
    public var name = "Untitled"
    public var composer = ""
    public var charter = ""
    public var level = ""
    public var chartID = ""
    public var song = ""
    public var background = ""
    public var offsetMilliseconds = 0
    public var rpeVersion = 123
    public var bpmChanges: [BPMChange]
    public var judgeLines: [JudgeLine]
    public var rawRoot: [String: JSONValue] = [:]
    public var rawMeta: [String: JSONValue] = [:]
    public private(set) var revision: UInt64 = 0

    public init() {
        bpmChanges = [BPMChange()]
        judgeLines = [JudgeLine()]
    }

    private init(empty _: Bool) {
        bpmChanges = []
        judgeLines = []
    }

    public init(data: Data) throws {
        let root: JSONValue
        do {
            root = try JSONValue.parse(data)
        } catch {
            throw ChartError.invalidJSON(error.localizedDescription)
        }
        try self.init(json: root)
    }

    public init(json: JSONValue) throws {
        guard let root = json.objectValue else {
            throw ChartError.invalidJSON("Chart root must be an object")
        }
        self.init(empty: true)
        rawRoot = root.removing(["META", "BPMList", "judgeLineList"])
        if let meta = root.object("META") {
            rawMeta = meta.removing([
                "RPEVersion", "background", "charter", "composer", "id",
                "level", "name", "offset", "song",
            ])
            name = meta.string("name", default: "Untitled")
            composer = meta.string("composer")
            charter = meta.string("charter")
            level = meta.string("level")
            chartID = meta.string("id")
            song = meta.string("song")
            background = meta.string("background")
            offsetMilliseconds = meta.int("offset")
            rpeVersion = meta.int("RPEVersion", default: 123)
        }
        bpmChanges = try (root.array("BPMList") ?? []).compactMap { value in
            guard value.objectValue != nil else { return nil }
            return try BPMChange(json: value)
        }
        if bpmChanges.isEmpty { bpmChanges = [BPMChange()] }
        bpmChanges.sort { $0.startTime < $1.startTime }

        judgeLines = try (root.array("judgeLineList") ?? []).compactMap { value in
            guard value.objectValue != nil else { return nil }
            return try JudgeLine(json: value)
        }
        if judgeLines.isEmpty { judgeLines = [JudgeLine()] }
    }

    public var json: JSONValue {
        var root = rawRoot
        var meta = rawMeta
        meta["RPEVersion"] = .integer(Int64(rpeVersion))
        meta["background"] = .string(background)
        meta["charter"] = .string(charter)
        meta["composer"] = .string(composer)
        meta["id"] = .string(chartID)
        meta["level"] = .string(level)
        meta["name"] = .string(name)
        meta["offset"] = .integer(Int64(offsetMilliseconds))
        meta["song"] = .string(song)
        root["META"] = .object(meta)
        root["BPMList"] = .array(bpmChanges.sorted { $0.startTime < $1.startTime }.map(\.json))
        root["judgeLineList"] = .array(judgeLines.map(\.json))
        if root["judgeLineGroup"] == nil {
            root["judgeLineGroup"] = .array([.string("Default")])
        }
        return .object(root)
    }

    public func encoded(prettyPrinted: Bool = false) throws -> Data {
        try json.encoded(prettyPrinted: prettyPrinted)
    }

    public mutating func markEdited() {
        revision &+= 1
    }

    public mutating func sort() {
        bpmChanges.sort { $0.startTime < $1.startTime }
        for index in judgeLines.indices {
            judgeLines[index].notes.sort { $0.startTime < $1.startTime }
            for layerIndex in judgeLines[index].eventLayers.indices {
                for type in EventType.allCases {
                    judgeLines[index].eventLayers[layerIndex].events[type]?.sort {
                        $0.startTime < $1.startTime
                    }
                }
            }
        }
    }

    public func bpm(at beat: Double) -> Double {
        var current = 120.0
        for change in bpmChanges {
            if change.startTime.doubleValue > beat { break }
            current = validBPM(change.bpm, fallback: current)
        }
        if let first = bpmChanges.first, first.startTime.doubleValue > beat {
            current = validBPM(first.bpm, fallback: current)
        }
        return current
    }

    /// Converts chart beat to elapsed chart milliseconds, excluding META offset.
    public func milliseconds(atBeat target: Double) -> Int64 {
        let beat = max(0, target.isFinite ? target : 0)
        if beat == 0 { return 0 }
        var milliseconds = 0.0
        var cursorBeat = 0.0
        var currentBPM = bpm(at: 0)
        for change in bpmChanges {
            let changeBeat = max(0, change.startTime.doubleValue)
            if changeBeat <= cursorBeat {
                currentBPM = validBPM(change.bpm, fallback: currentBPM)
                continue
            }
            if changeBeat >= beat { break }
            milliseconds += (changeBeat - cursorBeat) * 60_000 / currentBPM
            cursorBeat = changeBeat
            currentBPM = validBPM(change.bpm, fallback: currentBPM)
        }
        milliseconds += (beat - cursorBeat) * 60_000 / currentBPM
        return Int64(max(0, milliseconds.rounded()))
    }

    /// Converts elapsed chart milliseconds, excluding META offset, into a beat.
    public func beat(atMilliseconds target: Int64) -> Double {
        var remaining = Double(max(0, target))
        var cursorBeat = 0.0
        var currentBPM = bpm(at: 0)
        for change in bpmChanges {
            let changeBeat = max(0, change.startTime.doubleValue)
            if changeBeat <= cursorBeat {
                currentBPM = validBPM(change.bpm, fallback: currentBPM)
                continue
            }
            let segmentMilliseconds = (changeBeat - cursorBeat) * 60_000 / currentBPM
            if remaining <= segmentMilliseconds {
                return cursorBeat + remaining * currentBPM / 60_000
            }
            remaining -= segmentMilliseconds
            cursorBeat = changeBeat
            currentBPM = validBPM(change.bpm, fallback: currentBPM)
        }
        return cursorBeat + remaining * currentBPM / 60_000
    }

    public func audioMilliseconds(atBeat beat: Double, packageOffset: Int64 = 0) -> Int64 {
        let combined = Self.saturatingAdd(Int64(offsetMilliseconds), packageOffset)
        return max(0, Self.saturatingAdd(milliseconds(atBeat: beat), combined))
    }

    public func beat(atAudioMilliseconds milliseconds: Int64, packageOffset: Int64 = 0) -> Double {
        let combined = Self.saturatingAdd(Int64(offsetMilliseconds), packageOffset)
        return beat(atMilliseconds: max(0, Self.saturatingSubtract(milliseconds, combined)))
    }

    public var totalNotes: Int {
        judgeLines.reduce(0) { $0 + $1.notes.count }
    }

    public var totalEvents: Int {
        judgeLines.reduce(0) { $0 + $1.eventCount }
    }

    public var finalBeat: Double {
        var result = 1.0
        for line in judgeLines {
            for note in line.notes {
                result = max(
                    result,
                    max(note.startTime.doubleValue, note.endTime.doubleValue)
                )
            }
            for layer in line.eventLayers {
                for values in layer.events.values {
                    for event in values {
                        result = max(
                            result,
                            max(event.startTime.doubleValue, event.endTime.doubleValue)
                        )
                    }
                }
            }
            for values in line.storyboard.events.values {
                for event in values {
                    result = max(
                        result,
                        max(event.startTime.doubleValue, event.endTime.doubleValue)
                    )
                }
            }
        }
        return result
    }

    private func validBPM(_ value: Double, fallback: Double) -> Double {
        value.isFinite && value > 0 ? value : fallback
    }

    private static func saturatingAdd(_ left: Int64, _ right: Int64) -> Int64 {
        let result = left.addingReportingOverflow(right)
        if !result.overflow { return result.partialValue }
        return right >= 0 ? .max : .min
    }

    private static func saturatingSubtract(_ left: Int64, _ right: Int64) -> Int64 {
        if right == .min { return left >= 0 ? .max : left - right }
        return saturatingAdd(left, -right)
    }
}

File: app/src/main/java/com/xpe/mobile/model/ChartDocument.java
package com.xpe.mobile.model;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Comparator;
import java.io.IOException;
import java.io.StringWriter;
import java.io.Writer;
import java.util.Iterator;
import java.util.List;

public final class ChartDocument {
    public String name = "Untitled";
    public String composer = "";
    public String charter = "";
    public String level = "";
    public String id = "";
    public String song = "";
    public String background = "";
    public int offsetMs;
    public int rpeVersion = 123;
    public final List<BpmChange> bpmChanges = new ArrayList<>();
    public final List<JudgeLine> judgeLines = new ArrayList<>();
    private JSONObject rawRoot;
    private JSONObject rawMeta;
    private long revision;

    public static ChartDocument fromJson(String json) throws JSONException {
        return fromParsedJson(new JSONObject(json));
    }

    public static ChartDocument fromParsedJson(JSONObject root) throws JSONException {
        if (root == null) throw new JSONException("Chart JSON root is missing");
        ChartDocument chart = new ChartDocument();
        chart.rawRoot = RawJson.unknownFields(root, "META", "BPMList", "judgeLineList");

        JSONObject meta = root.optJSONObject("META");
        if (meta != null) {
            chart.rawMeta = RawJson.unknownFields(meta,
                    "RPEVersion", "background", "charter", "composer", "id",
                    "level", "name", "offset", "song");
            chart.name = meta.optString("name", "Untitled");
            chart.composer = meta.optString("composer", "");
            chart.charter = meta.optString("charter", "");
            chart.level = meta.optString("level", "");
            chart.id = meta.optString("id", "");
            chart.song = meta.optString("song", "");
            chart.background = meta.optString("background", "");
            chart.offsetMs = meta.optInt("offset", 0);
            chart.rpeVersion = meta.optInt("RPEVersion", 123);
        }

        JSONArray bpmArray = root.optJSONArray("BPMList");
        if (bpmArray != null) {
            for (int index = 0; index < bpmArray.length(); index++) {
                JSONObject bpmObject = bpmArray.optJSONObject(index);
                if (bpmObject != null) chart.bpmChanges.add(BpmChange.fromJson(bpmObject));
            }
        }
        if (chart.bpmChanges.isEmpty()) chart.bpmChanges.add(new BpmChange());
        chart.sortBpm();

        JSONArray linesArray = root.optJSONArray("judgeLineList");
        if (linesArray != null) {
            for (int index = 0; index < linesArray.length(); index++) {
                JSONObject lineObject = linesArray.optJSONObject(index);
                if (lineObject != null) chart.judgeLines.add(JudgeLine.fromJson(lineObject));
            }
        }
        if (chart.judgeLines.isEmpty()) chart.judgeLines.add(new JudgeLine());
        return chart;
    }

    public String toJsonString() throws JSONException {
        StringWriter output = new StringWriter();
        try {
            writeJson(output);
        } catch (IOException impossible) {
            throw new JSONException(impossible.getMessage());
        }
        return output.toString();
    }

    /** Streams compact JSON so large charts never exist as a second full in-memory String. */
    public void writeJson(Writer output) throws IOException, JSONException {
        output.write('{');
        boolean hasField = false;
        boolean hasGroups = false;
        if (rawRoot != null) {
            Iterator<String> keys = rawRoot.keys();
            while (keys.hasNext()) {
                String key = keys.next();
                if ("META".equals(key) || "BPMList".equals(key)
                        || "judgeLineList".equals(key)) continue;
                if (hasField) output.write(',');
                hasField = true;
                JsonStreamWriter.value(output, key);
                output.write(':');
                JsonStreamWriter.value(output, rawRoot.get(key));
                if ("judgeLineGroup".equals(key)) hasGroups = true;
            }
        }

        if (hasField) output.write(',');
        JsonStreamWriter.value(output, "META");
        output.write(':');
        JSONObject meta = RawJson.shallowCopy(rawMeta);
        meta.put("RPEVersion", rpeVersion);
        meta.put("background", background);
        meta.put("charter", charter);
        meta.put("composer", composer);
        meta.put("id", id);
        meta.put("level", level);
        meta.put("name", name);
        meta.put("offset", offsetMs);
        meta.put("song", song);
        JsonStreamWriter.value(output, meta);

        sortBpm();
        output.write(',');
        JsonStreamWriter.value(output, "BPMList");
        output.write(':');
        output.write('[');
        for (int index = 0; index < bpmChanges.size(); index++) {
            if (index > 0) output.write(',');
            JsonStreamWriter.value(output, bpmChanges.get(index).toJson());
        }
        output.write(']');

        output.write(',');
        JsonStreamWriter.value(output, "judgeLineList");
        output.write(':');
        output.write('[');
        for (int index = 0; index < judgeLines.size(); index++) {
            if (index > 0) output.write(',');
            JsonStreamWriter.value(output, judgeLines.get(index).toJson());
        }
        output.write(']');
        if (!hasGroups) {
            output.write(',');
            JsonStreamWriter.value(output, "judgeLineGroup");
            output.write(':');
            JsonStreamWriter.value(output, new JSONArray().put("Default"));
        }
        output.write('}');
    }

    public long revision() {
        return revision;
    }

    public void markEdited() {
        revision++;
    }

    public double bpmAt(double beat) {
        double current = 120.0;
        for (BpmChange change : bpmChanges) {
            if (change.startTime.toDouble() > beat) break;
            current = validBpmOrFallback(change.bpm, current);
        }
        if (!bpmChanges.isEmpty() && bpmChanges.get(0).startTime.toDouble() > beat) {
            current = validBpmOrFallback(bpmChanges.get(0).bpm, current);
        }
        return current;
    }

    /** Converts chart beat to elapsed chart milliseconds, excluding META offset. */
    public long beatToMillis(double targetBeat) {
        double beat = Math.max(0.0, targetBeat);
        if (beat == 0.0) return 0L;

        double milliseconds = 0.0;
        double cursorBeat = 0.0;
        double currentBpm = bpmAt(0.0);
        for (BpmChange change : bpmChanges) {
            double changeBeat = Math.max(0.0, change.startTime.toDouble());
            if (changeBeat <= cursorBeat) {
                currentBpm = validBpmOrFallback(change.bpm, currentBpm);
                continue;
            }
            if (changeBeat >= beat) break;
            milliseconds += (changeBeat - cursorBeat) * 60000.0 / currentBpm;
            cursorBeat = changeBeat;
            currentBpm = validBpmOrFallback(change.bpm, currentBpm);
        }
        milliseconds += (beat - cursorBeat) * 60000.0 / currentBpm;
        return Math.max(0L, Math.round(milliseconds));
    }

    /** Converts elapsed chart milliseconds, excluding META offset, into a beat. */
    public double millisToBeat(long targetMillis) {
        double remaining = Math.max(0L, targetMillis);
        double cursorBeat = 0.0;
        double currentBpm = bpmAt(0.0);
        for (BpmChange change : bpmChanges) {
            double changeBeat = Math.max(0.0, change.startTime.toDouble());
            if (changeBeat <= cursorBeat) {
                currentBpm = validBpmOrFallback(change.bpm, currentBpm);
                continue;
            }
            double segmentMillis = (changeBeat - cursorBeat) * 60000.0 / currentBpm;
            if (remaining <= segmentMillis) {
                return cursorBeat + remaining * currentBpm / 60000.0;
            }
            remaining -= segmentMillis;
            cursorBeat = changeBeat;
            currentBpm = validBpmOrFallback(change.bpm, currentBpm);
        }
        return cursorBeat + remaining * currentBpm / 60000.0;
    }

    public long beatToAudioMillis(double beat) {
        return beatToAudioMillis(beat, 0L);
    }

    /** Converts a beat to audio time using both RPE META.offset and package info.yml offset. */
    public long beatToAudioMillis(double beat, long packageOffsetMs) {
        long combinedOffsetMs = saturatingAdd(offsetMs, packageOffsetMs);
        return Math.max(0L, saturatingAdd(beatToMillis(beat), combinedOffsetMs));
    }

    public double audioMillisToBeat(long audioMillis) {
        return audioMillisToBeat(audioMillis, 0L);
    }

    /** Inverse of {@link #beatToAudioMillis(double, long)}. */
    public double audioMillisToBeat(long audioMillis, long packageOffsetMs) {
        long combinedOffsetMs = saturatingAdd(offsetMs, packageOffsetMs);
        return millisToBeat(Math.max(0L, saturatingSubtract(audioMillis, combinedOffsetMs)));
    }

    public int totalNotes() {
        int total = 0;
        for (JudgeLine line : judgeLines) total += line.notes.size();
        return total;
    }

    public int totalEvents() {
        int total = 0;
        for (JudgeLine line : judgeLines) total += line.countEvents();
        return total;
    }

    private static double validBpmOrFallback(double value, double fallback) {
        return Double.isFinite(value) && value > 0.0 ? value : fallback;
    }

    private static long saturatingAdd(long left, long right) {
        if (right > 0L && left > Long.MAX_VALUE - right) return Long.MAX_VALUE;
        if (right < 0L && left < Long.MIN_VALUE - right) return Long.MIN_VALUE;
        return left + right;
    }

    private static long saturatingSubtract(long left, long right) {
        if (right == Long.MIN_VALUE) return left >= 0L ? Long.MAX_VALUE : left - right;
        return saturatingAdd(left, -right);
    }

    public void sortBpm() {
        bpmChanges.sort(Comparator.comparing(change -> change.startTime));
    }
}
